Add Coupons autocompleter and relevant configs

psealock opened the pull request:
Add Coupons to Advanced Filters following established patterns.

<img width="891" alt="screen shot 2018-09-18 at 1 17 39 pm" src="https://user-images.githubusercontent.com/1922453/45658801-4932be80-bb45-11e8-9473-1e6b23bcef1a.png">

## Additional changes
 
* Coupons config including a `getCouponLabelsById` function
* Coupons report page
* Coupons autocompleter
* Removed unused stylesheets on orders and products reports

## Test
1. `/wp-admin/admin.php?page=wc-admin#/analytics/coupons?filter=compare`
2. See basic filters appear on the new Coupons report, including a Compare card
3. Make sure you have a few coupons on your test site and search for them
4. Ensure the basic compare functionality is present, including URL updates and tags appearing on page refresh
5. Do the same with Advanced Filters at `/wp-admin/admin.php?page=wc-admin#/analytics/orders?filter=advanced` by adding Coupon Codes filter.

Fixes https://github.com/woocommerce/wc-admin/issues/342

psealock committed:
Coupons: add config, autocompleter

allendav commented on client/analytics/report/coupons/config.js:
```diff
@@ -0,0 +1,45 @@
+	const idList = queryString
+		.split( ',' )
+		.map( id => parseInt( id, 10 ) )
+		.filter( Boolean );
```
I really appreciate your use of filter( Boolean ) here to avoid falsy things showing up in the idList, e.g. in the event someone messes with the queryString.  Would it be a good idea to remove duplicates too?

allendav replied:
(It is noteworthy that even with a request containing duplicates, the REST API only returns one blob per unique ID)

allendav commented on client/analytics/report/products/config.js:
```diff
@@ -26,7 +28,9 @@ export function getCategoryLabelsById( queryString ) {
 	} );
-	return apiFetch( { path: `${ NAMESPACE }products/categories${ payload }` } );
+	return apiFetch( { path: `${ NAMESPACE }products/categories${ payload }` } ).then( data =>
+		data.map( product => ( { id: product.id, label: product.name } ) )
```
Should the mapping be categories into category id and labels, not product ones?

allendav replied:
I mean, on further reflection this does work - but the map currentValue argument should be `category` and not `product` methinks

allendav reviewed (approved):
Works well. Please update currentValue variable name from product to category for category map call before committing. Please also consider de-duping queryString (totally optional.)

ryelle commented on client/components/search/autocompleters/coupons.js:
```diff
@@ -0,0 +1,59 @@
+ */
+export default {
+	name: 'coupons',
+	className: 'woocommerce-search__product-result', // @TODO remove, fix?
```
This isn't currently styled, so you could change it to `'woocommerce-search__coupon-result'`, or omit it entirely. It's not a required option.

ryelle reviewed (commented):
Just a few small notes, overall working well 👍

psealock commented:
Thanks for the reviews @allendav @ryelle 

* updated arg name `product` -> `category`
* de-duped ids in `getIdsFromQuery`
* updated className in coupons autocompleter

@ryelle I moved all the `getLabel` functions to `client/lib/async-requests`. I like how `/analytics` tends towards being free from logic, including things like util functions. I figured `lib/async-requests` could be a good place to keep requests outside of store. I'm totally open to suggestions, however.

Since all those functions were in the same place, I took the opportunity to dry things up a bit in https://github.com/woocommerce/wc-admin/pull/419/commits/2be9301e83163169d44249f8918ea7cc91bdd816 so configs looks like this:

```js
getLabels: getRequestByIdString( NAMESPACE + 'products/categories', category => ( {
    id: category.id,
    label: category.name,
} ) ),
```

I'm open to suggestions here as well since it increases complexity in the name of less code.

psealock commented:
I'm going to merge this one so it doesn't get stale, but I'm still open to suggestions on naming/moving the `getLabels` functions @ryelle

ryelle commented:
I've been thinking about the function name but I think this does work best – and `client/lib/async-requests` is a good place too. Thanks for drying things up 👍
